ping/ping.c

```diff
@@ -557,6 +557,15 @@ main(int argc, char **argv)
 	int target_ai_family = hints.ai_family;
 	hints.ai_family = AF_UNSPEC;
 
+	unsigned char buf[sizeof(struct in6_addr)];
+	if (!strchr(target, '%') && sock6.socktype == SOCK_DGRAM &&
+		inet_pton(AF_INET6, target, buf) > 0 &&
+		(IN6_IS_ADDR_LINKLOCAL(buf) || IN6_IS_ADDR_MC_LINKLOCAL(buf))) {
+			error(0, 0, _(
+				"Warning: IPv6 link-local address on non-raw ICMP socket requires ifname or scope-id"
+				" => use address%%<ifname|scope-id>"));
```

Comment:
The warning is appreciated but not completely accurate:
```
root@vsid:/src/iputils# ./builddir/ping/ping -I eth0 fe80::1 -c1
./builddir/ping/ping: Warning: IPv6 link-local address on non-raw ICMP socket requires ifname or scope-id => use address%<ifname|scope-id>
./builddir/ping/ping: Warning: source address might be selected on device other than: eth0
PING fe80::1(fe80::1) from :: eth0: 56 data bytes
64 bytes from fe80::1%eth0: icmp_seq=1 ttl=64 time=0.090 ms

--- fe80::1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.090/0.090/0.090/0.000 ms
```
ie. it requires `%` **or** `-I`

Reply:
1) while this works on some of my x86_64 machine, it's a problem on ppc64le
2) not mixing `-I ifname` with % notation was the reason to not taking https://github.com/iputils/iputils/issues/398, right? Do we want to reconsider it? With fixing `-I` parsing in getopts it would save VRF issue. That PR fixes ppc64le issues to make sure `-I ifname` works everywhere, even `-I ifindex` works with it).

*  `-I ifname` works on x86_64 machine:
```
$ ip -6 a show vnet0
7: vnet0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master virbr0 state UNKNOWN group default qlen 1000
    inet6 fe80::fc54:ff:feeb:ac2f/64 scope link 
       valid_lft forever preferred_lft forever

$ ip -6 a show wlp2s0 
2: wlp2s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    inet6 fe80::5054:ff:febb:a3d4/64 scope link 
       valid_lft forever preferred_lft forever
    inet6 fe80::766a:125a:c587:ab18/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
```
*  `-I ifname` broken on ppc64le machine:
```
$ ip -6 a show eth0 
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    altname enp0s1
    inet6 XXXX:XXX:XXXX:XXXX:XX:XXX:XX:XX/64 scope global dynamic noprefixroute 
       valid_lft 11374sec preferred_lft 5974sec
    inet6 fe80::5054:ff:febb:a3d4/64 scope link 
       valid_lft forever preferred_lft forever

# global IPv6 hidden

$ ./builddir/ping/ping fe80::5054:ff:febb:a3d4 -I eth0
./builddir/ping/ping: Warning: source address might be selected on device other than: eth0
PING fe80::5054:ff:febb:a3d4(fe80::5054:ff:febb:a3d4) from :: eth0: 56 data bytes

--- fe80::5054:ff:febb:a3d4 ping statistics ---
3 packets transmitted, 0 received, 100% packet loss, time 2062ms
```

Obviously % notation works (both ifname and ifindex):
```
$ ./builddir/ping/ping -c1 fe80::5054:ff:febb:a3d4%eth0
PING fe80::5054:ff:febb:a3d4%eth0(fe80::5054:ff:febb:a3d4%eth0) 56 data bytes
64 bytes from fe80::5054:ff:febb:a3d4%eth0: icmp_seq=1 ttl=64 time=0.085 ms

--- fe80::5054:ff:febb:a3d4%eth0 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.085/0.085/0.085/0.000 ms

$ ./builddir/ping/ping -c1 fe80::5054:ff:febb:a3d4%2
PING fe80::5054:ff:febb:a3d4%2(fe80::5054:ff:febb:a3d4%eth0) 56 data bytes
64 bytes from fe80::5054:ff:febb:a3d4%eth0: icmp_seq=1 ttl=64 time=0.073 ms

--- fe80::5054:ff:febb:a3d4%2 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.073/0.073/0.073/0.000 ms

```
I'm not sure whether it's ppc64le issue or issue with global IPv6 address affecting it. But deleting the global IPv6 address does not help.

I also added another iface on that ppc64le machine to test it, but it does not help
```
3: ltp_ns_veth2@if4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 state UP qlen 1000
    inet6 fd00:1:1:1::2/64 scope global nodad 
       valid_lft forever preferred_lft forever
    inet6 fe80::98cd:89ff:fee9:23b4/64 scope link 
       valid_lft forever preferred_lft forever

$ ./builddir/ping/ping fe80::98cd:89ff:fee9:23b4 -I ltp_ns_veth2
./builddir/ping/ping: Warning: source address might be selected on device other than: ltp_ns_veth2
PING fe80::98cd:89ff:fee9:23b4(fe80::98cd:89ff:fee9:23b4) from :: ltp_ns_veth2: 56 data bytes

--- fe80::98cd:89ff:fee9:23b4 ping statistics ---
8 packets transmitted, 0 received, 100% packet loss, time 7282ms
```
Obviously there is nothing ppc64le specific in iputils, IMHO the issue is not caused by network setup. Thus it's probably some ppc64le specific kernel network internals (or glibc), but unlikely a bug. Maybe I should ask on linuxppc-dev and netdev ML.

Reply:
> * `-I ifname` broken on ppc64le machine:

Interesting. I was curious so I tried using qemu. `-I` worked for me:
```
root@debian:~# ping fe80::1
PING fe80::1(fe80::1) 56 data bytes
^C
--- fe80::1 ping statistics ---
3 packets transmitted, 0 received, 100% packet loss, time 2041ms

root@debian:~# ping -I enp0s2 fe80::1 -c1
ping: Warning: source address might be selected on device other than: enp0s2
PING fe80::1(fe80::1) from :: enp0s2: 56 data bytes
64 bytes from fe80::1%enp0s2: icmp_seq=1 ttl=64 time=0.412 ms

--- fe80::1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.412/0.412/0.412/0.000 ms
root@debian:~# uname -m
ppc64le
root@debian:~# lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 11 (bullseye)
Release:        11
Codename:       bullseye
root@debian:~# ping -V
ping from iputils 20210202
```

Reply:
@gobenji Thanks! Maybe it's openSUSE HW specific. Could you please post also kernel version (`uname -r`). And even a used config (or which config did you use)?

Reply:
I simply installed Debian Bullseye and changed the sysctl to enable ping sockets, no other special configuration:

```
root@debian:~# sysctl net.ipv4.ping_group_range
net.ipv4.ping_group_range = 0   10000
root@debian:~# uname -a
Linux debian 5.10.0-13-powerpc64le #1 SMP Debian 5.10.106-1 (2022-03-17) ppc64le GNU/Linux
root@debian:~# ip -6 addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 state UNKNOWN qlen 1000
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: enp0s2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 state UP qlen 1000
    inet6 fd0e:224b:829:4050:5054:ff:feb7:5a00/64 scope global dynamic mngtmpaddr
       valid_lft 86111sec preferred_lft 14111sec
    inet6 fe80::5054:ff:feb7:5a00/64 scope link
       valid_lft forever preferred_lft forever
root@debian:~# ip -6 rou
::1 dev lo proto kernel metric 256 pref medium
fd0e:224b:829:4050::/64 dev enp0s2 proto kernel metric 256 expires 86105sec pref medium
fe80::/64 dev enp0s2 proto kernel metric 256 pref medium
default via fe80::8885:5803:fdbd:c285 dev enp0s2 proto ra metric 1024 expires 1505sec hoplimit 64 pref medium
root@debian:~# ping -I enp0s2 fe80::1 -c1
ping: Warning: source address might be selected on device other than: enp0s2
PING fe80::1(fe80::1) from :: enp0s2: 56 data bytes
64 bytes from fe80::1%enp0s2: icmp_seq=1 ttl=64 time=4.23 ms

--- fe80::1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 4.229/4.229/4.229/0.000 ms
```

Reply:
After long research, given that `-Iifname` is not going to be fixed for 4.9 LTS branch (https://lore.kernel.org/netdev/YppqNtTmqjeR5cZV@pevik/) or people may use unfixed kernels I'd keep suggesting only `address%<ifname|scope-id>`. But I'll put some info to the man page.